---
title: "Basic Serut Flow - SuperAgers"
output: html_notebook
---
---
title: "5% DB"
output: html_notebook
---

#Installation of the SeuratDisk & SeuratData:

```{r (Optional) installations}
#if (!requireNamespace("remotes", quietly = TRUE)) {
#  install.packages("remotes")
#}
#remotes::install_github("mojaveazure/seurat-disk")
#remotes::install_github('satijalab/seurat-data')
#install.packages("cli")
#install.packages("crayon")
#install.packages("hdf5r")
#install.packages("R6")
#install.packages("Matrix")
#install.packages("rlang")
#install.packages("withr")
#install.packages("SeuratData")
#install.packages("withr")
install.packages("hrbrthemes")

```


```{r Libreries} 

library(Seurat)
library(withr)
library(Seurat)
library(SeuratDisk)
library(SeuratData)
library(ggplot2)
library(gridExtra)
library(ggplot2)
library(dplyr)
library(plotly)
library(hrbrthemes)

```

Loading the object:
```{r} 
five_prec <- LoadH5Seurat("/Users/itamar_shahar/Downloads/objects/SuperAgersRandomSubset0.05.h5seurat")
five_prec
#with_pdf(file.path("/Users/itamar_shahar/Library/CloudStorage/GoogleDrive-itamar.shahar2@mail.huji.ac.il/My Drive/University/General/3rd_year_project/3rd_year_project/plots", "test.pdf"), width = 7, height = 5)
#ten_prec <- LoadH5Seurat("/Users/itamar_shahar/Downloads/objects/SuperAgersRandomSubset0.1.h5seurat")



with_png(file.path(tempdir(), "test.png"), width = 800, height = 600,
plot(runif(5))
)
```


#Basis look at the parameters:
Which genders are there: 



```{r Which genders are there}
unique(five_prec$Gender)
```           

Which Ages are there: 
```{r Which Ages are there }
unique(five_prec$Age)
```
Which orig.ident are there: 
```{r Which orig.ident are there}
unique(five_prec$orig.ident)
```
Which sample_number are there: 
```{r Which sample_number are there}
sort(unique(five_prec$sample_number))
```

Which SampleID are there: 
```{r Which SampleID are there:}

unique_sample_ids <- unique(five_prec$SampleID)
unique_sample_ids# Get unique SampleID values
sorted_unique_sample_ids <- sort(unique_sample_ids)  # Sort the unique values
count <- length(sorted_unique_sample_ids)  # Count the number of sorted unique values

# Print the count of sorted unique SampleID values
message("Number of different people: ", count)

```

Which Diagnosis are there: 
```{r Which diagnosis are there}
unique(five_prec$Diagnosis)
sum(five_prec$Diagnosis == "HA")

```
Which pmi are there: 
```{r Which pmi are there}
sort(unique(five_prec$pmi))

```
Which BRAAK are there: 
```{r Which BRAAK are there}
unique(five_prec$BRAAK)

```


Which Overall.CAA.Score are there: 
```{r Which Overall.CAA.Score are there}
unique(five_prec$Overall.CAA.Score)
```

Which C.score  are there: 
```{r Which C.score  are there}
unique(five_prec$C.score)
```
Which Thal  are there: 
```{r Which Thal  are there}
unique(five_prec$Thal)
```


Which Lewy.Body.Disease  are there: 
```{r Which Lewy.Body.Disease  are there}
unique(five_prec$Lewy.Body.Disease)

```
Which APOE.genotype   are there: 
```{r Which APOE.genotype   are there}
unique(five_prec$APOE.Genotype)

```
#How many cells per donor? 
```{r }
#result <- five_prec@meta.data %>%
#  group_by(sample_number) %>%
#  summarize(number_of_cells = n())
#?group_by
print("Celles per Donor")
#print(result)
```

```{r  adding the SampleID + Diagnosis col }

five_prec@meta.data["SampleID_Diagnosis"] <- paste(five_prec$SampleID, five_prec$Diagnosis, sep = "_")
five_prec[["SampleID_Diagnosis"]] <- paste(five_prec$SampleID, five_prec$Diagnosis, sep = "_")
#head(five_prec@meta.data, 5)
```


# QC 

```{r setup, include=FALSE}
#head(five_prec@meta.data, 5)
five_prec[["percent.mt"]] <- PercentageFeatureSet(five_prec, pattern = "^MT-", assay = "RNA")
summary(five_prec@meta.data$percent.mt)
tail(sort(five_prec@meta.data$percent.mt), 400)
#head(five_prec@assays)#["CTCATCATCTAAGGAG-7"]#$counts
```

For this stage there are multiple of ways to look for those outlayers: 
We can use the `group.by` param and set it to: `Diagnosis` or `SampleID`. 
In this case we choose to use `group.by = Diagnosis`
```{r qc2, fig.height=7, fig.width=13}
#Visualize QC metrics as a violin plot
VlnPlot(five_prec, 
        features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), 
        ncol = 3,
        #group.by = "SampleID")
        group.by = "Diagnosis")#,
        #idents = 1:5)
sort(unique(five_prec@active.ident))
# FeatureScatter is typically used to visualize feature-feature relationships, but can be used for anything calculated by the object, i.e. columns in object metadata, PC scores etc.
```

```{r}
plot1 <- FeatureScatter(five_prec, 
                        feature1 = "nCount_RNA", 
                        feature2 = "percent.mt", 
                        #group.by = "Diagnosis",
                        group.by = "SampleID")
            
plot2 <- FeatureScatter(five_prec,
                        feature1 = "nCount_RNA",
                        feature2 = "nFeature_RNA",
                        #group.by = "Diagnosis",
                        group.by = "SampleID")

plot1 + plot2
#summary(five_prec@meta.data$percent.mt)
#five_prec <- subset(five_prec, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5)
```

```{r}
five_prec <- subset(five_prec, subset = nFeature_RNA > 200 & nFeature_RNA < 13000 & percent.mt < 10)
```
## Normalizing the data

```{r normalize}
five_prec <- SCTransform(five_prec, 
                        conserve.memory = TRUE,
                         ) 
                         #normalization.method = "LogNormalize",
                         #scale.factor = 10000)
?SCTransform
```
## Identification of highly variable features (feature selection)

```{r var_features, fig.height=5, fig.width=11}

top10 <- head(VariableFeatures(five_prec), 10)
top10
#?VariableFeaturePlot
# plot variable features with and without labels
plot1 <- VariableFeaturePlot(five_prec)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot1 + plot2
```

***

## Scaling the data
this stage is relevant if we didn't the `SCTransform()`  

```{r regress, fig.height=7, fig.width=11, results='hide'}
#five_prec <- ScaleData(five_prec)
```

## Perform linear dimensional reduction

Next we perform PCA on the scaled data. By default, only the previously determined variable features are used as input, but can be defined using `features` argument if you wish to choose a different subset.

```{r pca,results='hide'}
five_prec <- RunPCA(five_prec, features = VariableFeatures(object = five_prec))
```

Seurat provides several useful ways of visualizing both cells and features that define the PCA, including `VizDimReduction()`, `DimPlot()`, and `DimHeatmap()`

```{r pca_viz, message=TRUE}
# Examine and visualize PCA results a few different ways
print(five_prec[['pca']], dims = 1:5, nfeatures = 5)
VizDimLoadings(five_prec, dims = 1:2, reduction = 'pca')

```
```{r pca_viz, message=TRUE}

DimPlot(five_prec, 
        reduction = 'pca',
        #group.by = "Diagnosis",
        group.by = "SampleID")
```

In particular `DimHeatmap()` allows for easy exploration of the primary sources of heterogeneity in a dataset, and can be useful when trying to decide which PCs to include for further downstream analyses. Both cells and features are ordered according to their PCA scores. Setting `cells` to a number plots the 'extreme' cells on both ends of the spectrum, which dramatically speeds plotting for large datasets. Though clearly a supervised analysis, we find this to be a valuable tool for exploring correlated feature sets.

```{r single-heatmap}
DimHeatmap(five_prec, dims = 1, cells = 500, balanced = TRUE)
```

```{r multi-heatmap, fig.height=15, fig.width=9}
DimHeatmap(five_prec, dims = 1:15, cells = 500, balanced = TRUE)
```

## Determine the 'dimensionality' of the dataset



```{r jackstraw, fig.height=6, fig.width=10}
# NOTE: This process can take a long time for big datasets, comment out for expediency. More approximate techniques such as those implemented in ElbowPlot() can be used to reduce computation time
#five_prec <- JackStraw(five_prec, num.replicate = 100)
#five_prec <- ScoreJackStraw(five_prec, dims = 1:20)
```


```{r jsplots, fig.height=6, fig.width=10}
#JackStrawPlot(five_prec, dims = 1:15)
```

An alternative heuristic method generates an 'Elbow plot': a ranking of principle components based on the percentage of variance explained by each one (`ElbowPlot()` function). In this example, we can observe an 'elbow' around PC9-10, suggesting that the majority of true signal is captured in the first 10 PCs. 

```{r elbow_plot, fig.height=6, fig.width=10}
ElbowPlot(five_prec)
```

## Cluster the cells

```{r cluster, fig.height=5, fig.width=7}
five_prec <- FindNeighbors(five_prec, dims = 1:15)
five_prec <- FindClusters(five_prec, resolution = 0.5)

# Look at cluster IDs of the first 5 cells
head(Idents(five_prec), 5)
```

***

## Run non-linear dimensional reduction (UMAP/tSNE)

Seurat offers several non-linear dimensional reduction techniques, such as tSNE and UMAP, to visualize and explore these datasets. The goal of these algorithms is to learn the underlying manifold of the data in order to place similar cells together in low-dimensional space. Cells within the graph-based clusters determined above should co-localize on these dimension reduction plots. As input to the UMAP and tSNE, we suggest using the same PCs as input to the clustering analysis.

```{r tsne, fig.height=5, fig.width=7}
# If you haven't installed UMAP, you can do so via reticulate::py_install(packages = "umap-learn")
five_prec <- RunUMAP(five_prec, dims = 1:15)
```

```{r tsneplot, fig.height=5, fig.width=7}
# note that you can set `label = TRUE` or use the LabelClusters function to help label individual clusters

DimPlot(five_prec,
        reduction = 'umap', 
        #group.by ="Diagnosis",
        #group.by ="SampleID",
        )
?DimPlot
```

You can save the object at this point so that it can easily be loaded back in without having to rerun the computationally intensive steps performed above, or easily shared with collaborators.

```{r saveobject, eval=FALSE}
#saveRDS(pbmc, file = "../output/pbmc_tutorial.rds")
```



## Finding differentially expressed features (cluster biomarkers)


```{r markers1, fig.height=8, fig.width=15}
# find all markers of cluster 2
?FindMarkers


cluster2.markers <- FindMarkers(five_prec, 
                                ident.1 = 0,
                                min.pct = 0.25)
head(cluster2.markers, n = 5)


cluster2.markers["pct1_div_pct2"] <- abs(cluster2.markers$pct.1 - cluster2.markers$pct.2)
head(cluster2.markers, n = 5)

sorted_cluster2.markers <- cluster2.markers[order(cluster2.markers$pct1_div_pct2,decreasing = TRUE), ]
head(sorted_cluster2.markers, n = 40)

saveRDS(cell_markers, "cells_marker")
cell_markers <- list(
  "Neuroepithelial_cells" = c("SOX2", "Notch1", "HES1", "HES3", "E-cadherin",
                              "occludin", "NES", "NOTCH1", "OCLN"),
  "Radial_glia" = c("Vimentin", "nestin", "PAX6", "HES1", "HES5", "GFAP", "GLAST", "BLBP", "TN-C", "N-cadherin", "SOX2", "NES"),
  "Intermediate_progenitors" = c("TBR2", "MASH1/Ascl1"),
  "Immature_neurons" = c("DCX", "beta_III_tubulin", "NEUROD1", "TBR1", "STMN1"),
  "Oligodendrocyte_precursor_cells" = c("PDGFRA", "CSPG4"),
  "Mature_oligodendrocytes" = c("Olig_1", "olig_2", "olig_3", "MBP", "OSP", "MOG", "SOX10"),
  "Schwann_cells" = c("MPZ", "NCAM1", "GAP43","S100A1", "S100A2", "S100A3", "NGFR"),
  "Astrocytes" = c("GFAP", "SLC1A3", "SLC1A2", "glutamine_synthetase", "S100B",
                   "ALDH1L1"),
  "Microglia" = c("TMEM119", "CD11b", "CD45", "Iba1", "CX3CR1", "F4/80", "CD68", "CD40"
                  ,"PTPRC", "ITGAM", "AIF1", "EMR1", "ADGRE1"),
  "Mature_neurons" = c("NeuN", "MAP2", "160_kDa_neurofilament_medium", "200kDa_neurofilament_heavy", "synaptophysin", "PSD95"
                       ,"RBFOX3", "DLG4", "SYP"),
  
  
  
  
  "Glutamatergic_neurons" = c("VGLUT1", "VGLUT2", "NMDAR1", "NMDAR2B", "glutaminase", "glutamine_synthetase", "SLC17A7", "SLC17A6", "GRIN1", "GRIN2B", "SLC6A1"),
  
  "GABAergic_neurons" = c("GABA_transporter_1", "GABAB_receptors_1_and_2", 
                          "GAD65", "GAD67", "GAD2", "GAD1"),
  
  "Dopaminergic_neurons" = c("Tyrosine_hydroxylase", "dopamine_transporter", "FOXA2", "GIRK2", "Nurr1", "LMX1B"),
  "Serotonergic_neurons" = c("Tryptophan_hydroxylase", "serotonin_transporter", "Pet1"),
  "Cholinergic_neurons" = c("Choline_acetyltransferase", "vesicular_acetylcholine_transporter", "acetylcholinesterase"),
  "vacular" = c("PECAM1", "PTPRC", "ICAM4-AS1","LYVE1", "TEK","VCAM1", "CDH1", "VWF")
)

```

```{r}
Oligodendrocyte_precursor_cells_plot <- DotPlot(object = five_prec, 
        features = cell_markers["Oligodendrocyte_precursor_cells"], 
        scale.min = 0,
        scale.max = 100, 
)
Schwann_cells_plot <- DotPlot(object = five_prec, 
        features = cell_markers["Schwann_cells"],
        scale.min = 0,
        scale.max = 100
)
Astrocytes_plot <- DotPlot(object = five_prec, 
        features = cell_markers["Astrocytes"],
        scale.min = 0,
        scale.max = 100
)
Microglia_plot <- DotPlot(object = five_prec, 
        features = cell_markers["Microglia"],
        scale.min = 0,
        scale.max = 100
)
Mature_neurons_plot <- DotPlot(object = five_prec, 
        features = cell_markers["Mature_neurons"],
        scale.min = 0,
        scale.max = 100
)
Glutamatergic_neurons_plot <- DotPlot(object = five_prec, 
        features = cell_markers["Glutamatergic_neurons"],
        scale.min = 0,
        scale.max = 100
        )
GABAergic_neurons_plot <- DotPlot(object = five_prec, 
        features = cell_markers["GABAergic_neurons"],
        scale.min = 0,
        scale.max = 100
        )
vacular_plot <- DotPlot(object = five_prec, 
        features = cell_markers["vacular"],
        scale.min = 0,
        scale.max = 100
        )
vacular_plot


```
```{r}
?grid.arrange
#gridPlot(gridPlot)
g <- grid.arrange(Oligodendrocyte_precursor_cells_plot, 
         #Schwann_cells_plot, 
         Astrocytes_plot, 
         Microglia_plot, 
         Mature_neurons_plot, 
         Glutamatergic_neurons_plot, 
         GABAergic_neurons_plot,
         vacular_plot,
         nrow=3, ncol=2)
#g <- ggplotly(g)
#g
ggsave("tmp.pdf", 
       g, 
       width = 35, 
       height = 15, limitsize = FALSE)
```
### So we can say:
-> that cluster 1 & 14 contains Microglia
-> that cluster 7 contains Oligodendrocyte_precursor_cells

```{r}
head(cluster2.markers, n = 5)
# find all markers distinguishing cluster 5 from clusters 0 and 3
cluster5.markers <- FindMarkers(five_prec, ident.1 = 5, ident.2 = c(0, 3), min.pct = 0.25)
head(cluster5.markers, n = 5)




# find markers for every cluster compared to all remaining cells, report only the positive ones
pbmc.markers <- FindAllMarkers(pbmc, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
pbmc.markers %>% group_by(cluster) %>% slice_max(n = 2, order_by = avg_log2FC)
```

Seurat has several tests for differential expression which can be set with the test.use parameter (see our [DE vignette](de_vignette.html) for details). For example, the ROC test returns the 'classification power' for any individual marker (ranging from 0 - random, to 1 - perfect).

```{r markersroc, fig.height=8, fig.width=15}
levels(five_prec)
cluster0.markers <- FindMarkers(pbmc, ident.1 = 0, logfc.threshold = 0.25, test.use = "roc", only.pos = TRUE)
```

We include several tools for visualizing marker expression. `VlnPlot()` (shows expression probability distributions across clusters), and `FeaturePlot()` (visualizes feature expression on a tSNE or PCA plot) are our most commonly used visualizations. We also suggest exploring `RidgePlot()`, `CellScatter()`, and `DotPlot()` as additional methods to view your dataset.

```{r markerplots, fig.height=10, fig.width=15}
VlnPlot(pbmc, features = c("MS4A1", "CD79A"))
# you can plot raw counts as well
VlnPlot(pbmc, features = c("NKG7", "PF4"), slot = 'counts', log = TRUE)
FeaturePlot(pbmc, features = c("MS4A1", "GNLY", "CD3E", "CD14", "FCER1A", "FCGR3A", "LYZ", "PPBP", "CD8A"))
```

```{r clusterHeatmap, fig.height=8, fig.width=15}
pbmc.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_log2FC) -> top10
DoHeatmap(pbmc, features = top10$gene) + NoLegend()
```

```{r labelplot, fig.height=5, fig.width=9}
new.cluster.ids <- c("Naive CD4 T", "CD14+ Mono", "Memory CD4 T", "B", "CD8 T", "FCGR3A+ Mono", "NK", "DC", "Platelet")
names(new.cluster.ids) <- levels(pbmc)
pbmc <- RenameIdents(pbmc, new.cluster.ids)
DimPlot(pbmc, reduction = 'umap', label = TRUE, pt.size = 0.5) + NoLegend()
```

```{r save.img, include=FALSE}
library(ggplot2)
plot <- DimPlot(pbmc, reduction = "umap", label = TRUE, label.size = 4.5) + xlab("UMAP 1") + ylab("UMAP 2") + 
  theme(axis.title = element_text(size = 18), legend.text = element_text(size = 18)) + 
  guides(colour = guide_legend(override.aes = list(size = 10)))
ggsave(filename = "../output/images/pbmc3k_umap.jpg", height = 7, width = 12, plot = plot, quality = 50)
```

```{r save.rds, eval=FALSE}
saveRDS(pbmc, file = "../output/pbmc3k_final.rds")
```

```{r save2, include = FALSE}
saveRDS(pbmc, file = "../data/pbmc3k_final.rds")
```

```{r save.times, include = FALSE}
write.csv(x = t(as.data.frame(all_times)), file = "../output/timings/pbmc3k_tutorial_times.csv")
```


```{r}
sessionInfo()
```
</details>






